### ISSUE TITLE
Display the nearest data of different series in one tooltip.

### ISSUE THREAD

**Issue by RomanOlegovich** [apache/echarts#15488]
### What problem does this feature solve?
Is there a way to show the nearest data of different series in one tooltip?

Several series with a common x-axis are displayed on the chart.
Although the time between values may differ, I would like to display a general hint if the time difference is no more than a few seconds.
There can be about 8 series on the chart, a series of 30k - 100k points long, it is large dataset to dynamic search nearest data on tooltip formatter, also dynamic search is very slow. Store nearest data for each point is bad way also it's slow too.

Chart options:
The x-axis is time, the y-axis is value. 

the x-axis is time, the y-axis is the value.
Example:
```
Series 1: [{x: 1 ms}, {x: 3 ms}, {x: 5 ms}]
Series 2: [{x: 0 ms}, {x: 2 ms}, {x: 6 ms}]
```

### What does the proposed API look like?
May be

```
tooltip: {
    nearest: true,
    formatter: (params: Object|Array, nearest: Object|Array) =>
}
```

**Comment by Ovilia** [comment]
Are you using `'axis'` for [trigger](https://echarts.apache.org/en/option.html#tooltip.trigger)? Does [snap](https://echarts.apache.org/en/option.html#tooltip.axisPointer.snap) work for you? It may probably show series with the exact x value rather than nearby data of all series as you expected. I'm not sure if this will confuse the user to think that all series have the exact x value but in fact not?

**Comment by RomanOlegovich** [comment]
Hi @Ovilia . 
Yes I use 'axis' trigger.
Default value of snap is true, then I use snap. But it doesn't work for me. Can snap threshold be configured to show nearby data of all series?

**Comment by RomanOlegovich** [comment]
Is it possible with current library implementation?

**Comment by Ovilia** [comment]
No, it is not supported yet. Because we believe this will mislead the user to believe the second series also share the x value of the tooltip position.
If you need to implement this, you can get the x value with event and use a for loop to calculate the nearest value in each series and set to the tooltip formatter.

**Comment by Ovilia** [comment]
We could consider providing an extra option to enabled this new feature. To be discussed later.

**Comment by RomanOlegovich** [comment]
Also exist issue, after sampling (I use 'lttb', but for other this actual too) all series with same x-axis will not display on single tooltip.
Should I open new issue or this is connected with current?
![issue_tooltip_sampling_1](https://user-images.githubusercontent.com/46935437/129509373-8be3c110-56c8-454d-a590-3d5df59169b6.png)
![issue_tooltip_sampling_2](https://user-images.githubusercontent.com/46935437/129509376-c0a8a936-9da7-435c-8d64-d823cbcc7260.png)

**Comment by RomanOlegovich** [comment]
@Ovilia Also issue exists: after sampling not all series with same x-axis will display on single tooltip.
For this case configuration of snap threshold is required.
Should I open new issue or this is connected with current?

![issue_tooltip_sampling_1](https://user-images.githubusercontent.com/46935437/129512463-44748b46-fd43-403a-83fc-bdced75b21cc.png)
![issue_tooltip_sampling_2](https://user-images.githubusercontent.com/46935437/129512465-21d8bb39-2229-42ad-99eb-ee2c3eba5e46.png)

**Comment by Ovilia** [comment]
@Raphyyy Thanks for reporting. We are going to work on `axisPointer.snap` in the near future.

**Comment by Raphyyy** [comment]
Nice thanks!
I think the actual implementation of `axisPointer.snap` if `true` is to magnet the axisPointer to the nearest point, no matter the series.
The approach here is a bit different. The purpose is not to magnet axisPointer to the nearest point, it's to read all series' nearest point to get them in the tooltip, so I don't know if it is related to `axisPointer.snap` at the end.

**Comment by RndRss** [comment]
+1 for this

**Comment by petrjahoda** [comment]
+1 for this too

**Comment by Wintermute79** [comment]
+1 continuous reading of all series values at any x point would be nice - even if it's interpolated when there is no real value for a series at this x.

**Comment by Methodician** [comment]
+1 to this too! I just wrestled with this issue for days-worth of "spare time" until I finally generated some test data to realize what was causing my tooltip to not display as intended. I'm also using echarts to show time series from sensors, which never emit data at the exact same time.

I would like this feature very much (forcing tooltip to display all series to the nearest timeslot on hover).

If library managers are so concerned with misleading the viewer, one possibility would be to also display the timestamp for each series on the tooltip, instead of just the one. However, I would be happy if the tooltip just displayed the time where it was triggered, or the time of the first series it reads, or just about anyting.

**Comment by Methodician** [comment]
Found a workable work-around for folks in a similar situation to me...

My sensors emit data every 5 minutes, so I'm rounding the time to the nearest 5 minutes and it works like a charm.
I map through the readouts for each sensor to create my series. Here is some example code to get you started:

`
 const moduleSeries = modulesWithReadouts.map(({ name, readouts }) => {
        const series: SeriesOption = {
          name,
          type: 'line',
          data: readouts.map((readout: any) => {
            const coeff = 1000 * 60 * 5; // 5 minutes
            return [
              new Date(Math.round(readout.timestamp / coeff) * coeff),
              readout.temperature,
            ];
          }),
        };
        return series;
      });
`

**Comment by l1b3r** [comment]
The workaround from @Methodician might be viable in cases when you know the sampling frequency beforehand. However, in more general cases when your data is of various sampling rate it unfortunately does not work. Some datasets don't have a data point that would round to a specific point on the `x` axis while other datasets can round down to the same `x` multiple times. 

See the screenshot for an example of such scenario.
![image](https://user-images.githubusercontent.com/6207227/154655475-57498658-fc6b-4187-8fc8-fbbac6998165.png)


So, +1 from me for being able to snap to the nearest point within a given margin of error.

**Comment by leomerel** [comment]
For anyone who might be interested, I found a workaround that works for me. 
Here's the link to the stackoverflow answer : https://stackoverflow.com/questions/55238820/how-can-i-show-all-series-in-tooltip-when-series-does-not-have-same-time-values/72026277#72026277
![image](https://user-images.githubusercontent.com/41743205/165493595-99370566-099b-4f76-b0fa-d3969575e1b1.png)

**Comment by Wintermute79** [comment]
> It could also display the value that we see along the line (not an actual point then), by modifying the function.

How? that is the actual question. You did not provide that function code which gives you 23.9 from 2nd image above.

**Comment by leomerel** [comment]
> How? that is the actual question. You did not provide that function code which gives you 23.9 from 2nd image above.

I just added the code I wrote to achive that result in the stackoverflow answer. Hope it'll help !

**Comment by Wintermute79** [comment]
> > How? that is the actual question. You did not provide that function code which gives you 23.9 from 2nd image above.
> 
> I just added the code I wrote to achive that result in the stackoverflow answer. Hope it'll help !

Thank you so much! That solves it. Upvoted on stackoverflow.

**Comment by tuyuribr** [comment]
+1 for this

**Comment by namiwa** [comment]
+1 for this

**Comment by StefanKX** [comment]
+1 for this

**Comment by xitice** [comment]
+1 for this

**Comment by amartin-discovery** [comment]
+1

**Comment by mistical2008** [comment]
+1000 For this. 
The library has this weird behavior for some years. When it has data points misaligned by X then not all points comes to the `tooltip.formatter`. Doc searching doesn't help. No success.

**My current case for example:**
software getting multiple sensors data at different time interval. It leads to having misaligned series data  x-values. Maybe [this solution](https://github.com/apache/echarts/issues/15488#issuecomment-1110890817) could be help, but business has requirement to display big data. So iteration over big data list can suffer performance a lot. Maybe anyone solve this in any other way? I'm drained by constant experimenting on this.

**Comment by vildantursic** [comment]
Hey @Ovilia any news on this?

**Comment by RyKilleen** [comment]
After spending some time porting my solution to Echarts, I'll say that this comes as a very big surprise. I had assumed tooltips would display the interpolated value at a given point in time, or provide similar configuration to achieve it.

I don't have the ability to control the sampling of some of my data, several temperature time series, and that's leading to a very negative experience regarding tooltips.

**Comment by anton6** [comment]
+1

**Comment by Muqeet1** [comment]
+1

**Comment by KavTV** [comment]
+1

**Comment by Parallaps** [comment]
Any updates?

**Comment by Samsonium** [comment]
+1

**Comment by rehni** [comment]
+1

**Comment by rehni** [comment]
actually, the behavior where you snap more values on hover and which you think is misleading is wide used by other chart libraries. i didnt think this could be a problem in echarts.

**Comment by Hackerberg43** [comment]
+1

**Comment by EmptyBucket** [comment]
+1

**Comment by Montinew** [comment]
I found a practical workaround that works reliably for me.

**Problem:**
With asynchronous time series (timestamps differ slightly), tooltip aggregation fails because exact x-value alignment is required.

**Workaround:**
Bind each series to its own `time` xAxis and enable `tooltip.smartFormatter`.
smartFormatter allows the tooltip to display a single, consolidated timestamp instead of showing a separate timestamp per series.
Additional xAxes can be hidden.

```
{
  "tooltip": { "smartFormatter": true },
  "xAxis": [
    { "type": "time" },
    { "type": "time", "show": false },
    { "type": "time", "show": false }
  ],
  "yAxis": { "type": "value" },
  "series": [
    { "name": "Series 1", "type": "line", "xAxisIndex": 0, "data": [[t1, v1]] },
    { "name": "Series 2", "type": "line", "xAxisIndex": 1, "data": [[t2, v2]] },
    { "name": "Series 3", "type": "line", "xAxisIndex": 2, "data": [[t3, v3]] }
  ]
}
```
<img width="1462" height="662" alt="Image" src="https://github.com/user-attachments/assets/6400009a-75ec-4515-9751-4c3ffff35b8e" />

**Comment by jkbgntr94** [comment]
+1
